README.md
# eoir
Puppeteer functions to retrieve data from eoir site

[api](https://github.com/tonybranfort/eoir/blob/master/docs/api.md)


CHANGELOG.md
# v1.0.1
- No functionality change; replace utils~getDiffs with module ldb-change getDiffs
# v1.0.0 
- Initial release




test/case.test.js
/** tests for eoir cases
**/ 

let should = require('should'); 
let fs = require('fs');

const pfns = require('../lib/pfns.js');
let eoirLogin = require('../lib/login.js'); 
let eoirCase = require('../lib/case.js'); 
let eoirCaseList = require('../lib/case-list.js'); 
const utils = require('../lib/utils');

let config;
getConfig(); 

function getConfig() {
  config = JSON.parse(fs.readFileSync('test/test-config.json', 'utf8'));
}

config.testCases.forEach(testCase => {
  testCase.chargingDocDate.date = toDate(testCase.chargingDocDate.date);
  testCase.proceedings.forEach((proceeding) => {
    proceeding.decisionDate.date = toDate(proceeding.decisionDate.date);
  });
  testCase.hearings.forEach((hearing) => {
    hearing.date.date = toDate(hearing.date.date);
  });
}); 

function toDate(d) {
  return d ? new Date(d) : d;
}


describe('eoir case',function(){
  this.timeout(50000); 

  let page; 

  it('should get a browser page and login user', function() {
    return pfns.getNewBrowserPage({headless: false, slowMo: 100})
    .then((p) => {
      page = p;
      return eoirLogin.login(page, config.loginCreds);
    });
  }); // end of it

  it('should go to a case', function() {
    let id; 

    return eoirCase.gotoCase(page, config.testCases[0].eoirId)
    .then(() => pfns.getEl(page, eoirCase.selectors.aNbr))
    .then((aNbrEl) => (aNbrEl === null).should.equal(false));
    // .then(() => page.waitFor(10000))
    // .then(() => 'abc'.should.equal('abc'));

  }); 

  config.testCases.forEach(validateCase);

  function validateCase(testCase) {
    it('should get the case detail data for ' + testCase.eoirId, () => {
      return eoirCase.getCaseData(page, testCase.eoirId)
      .then((caseObj) => {
        // console.log(caseObj);
        let d = utils.getDiffs(testCase, caseObj);
        if(d.length > 0 ) {console.log(d);}
        d.length.should.equal(0);
      });
    });
  }

});




test/pfns.test.js
/** tests for eoir login and navigation
**/

let should = require('should'); 
let fs = require('fs');

let eoirLogin = require('../lib/login.js'); 
let eoirHome = require('../lib/home'); 
const pfns = require('../lib/pfns.js');

let config;
getConfig(); 

function getConfig() {
  config = JSON.parse(fs.readFileSync('test/test-config.json', 'utf8'));
}

describe('eoir pfns',function(){
  this.timeout(50000); 

  let page; 

  it('should get a browser page and login user', function() {
    return pfns.getNewBrowserPage({headless: false, slowMo: 100})
    .then((p) => {
      page = p;
      return eoirLogin.login(page, config.loginCreds);
    });
  }); // end of it

  it('should retrieve an element within nested iframes', function() {
    let myCasesLink =  [
      'main_frame', 
      'main_iframe', 
      'a#TilesViewApplicationList1_appDataList_ctl00_appItemLink'];

    return eoirHome.gotoHome(page)
    .then(() => pfns.getEl(page, myCasesLink))
    .then((el) => {
      // console.log(el);
      return (el === null).should.equal(false);
      // return page.waitFor(5000);
    });

  });

});


test/util.test.js
/** tests for utils
**/ 

let should = require('should'); 
let utils = require('../lib/utils');

describe('utils',function(){

  describe('getDiffs', () => {
    it('should compare primitives', function() {
      let d = utils.getDiffs(4, 'a');
      d.length.should.equal(1);
      d[0].path.should.equal('');
      d[0].lObj.should.equal(4);
      d[0].rObj.should.equal('a');
    }); // end of it
    
    it('should determine diffs for simple 1 level objects', function() {
      let lo = {a: 1, b: 2, c: 5};
      let ro = {a: 2, b: 2};
      let d = utils.getDiffs(lo, ro);
      d.length.should.equal(2);
      d[0].path.should.equal('a');
      d[0].lObj.should.equal(1);
      d[0].rObj.should.equal(2);
      d[1].path.should.equal('c');
      d[1].lObj.should.equal(5);
      (d[1].rObj === undefined).should.equal(true);
    }); // end of it

    it('should ignore props in rObj that are not in lObj', function() {
      let lo = {a: 1};
      let ro = {a: 1, b: 2};
      let d = utils.getDiffs(lo, ro);
      d.length.should.equal(0);
    }); // end of it

    it('should get diffs for multi level objects', function() {
      let lo = {a: 1, b: {c: 2, d: 3, e: {f: 6}}};
      let ro = {a: 1, b: {c: 2, d: 5, e: {g: 6}}};
      let d = utils.getDiffs(lo, ro);
      d.length.should.equal(2);
      d[0].path.should.equal('b.d');
      d[0].lObj.should.equal(3);
      d[0].rObj.should.equal(5);
      d[1].path.should.equal('b.e.f');
      d[1].lObj.should.equal(6);
      (d[1].rObj === undefined).should.equal(true);
    }); // end of it

    it('should get diffs for nested arrays', function() {
      let lo = {a: 1, b: [{c: 6}, {c: 8}, {c: 10}]};
      let ro = {a: 1, b: [{c: 6}, {c: 9}]};
      let d = utils.getDiffs(lo, ro);
      d.length.should.equal(2);
      d[0].path.should.equal('b[1].c');
      d[0].lObj.should.equal(8);
      d[0].rObj.should.equal(9);
      d[1].path.should.equal('b[2]');
      d[1].lObj.c.should.equal(10);
      (d[1].rObj === undefined).should.equal(true);
    }); // end of it

    it('should compare dates', function() {
      let lo = {a: new Date('2017/01/01'), b: new Date('2018/01/01')};
      let ro = {a: new Date('2017/12/01'), b: new Date('2018/01/01')};
      let d = utils.getDiffs(lo, ro);
      d.length.should.equal(1);
      d[0].path.should.equal('a');
      d[0].lObj.should.equal(lo.a);
      d[0].rObj.should.equal(ro.a);
    }); // end of it
  
    it('should compare different type', function() {
      let lo = {a: new Date('2017/01/01'), b: new Date('2018/01/01')};
      let ro = {a: 'q', b: new Date('2018/01/01')};
      let d = utils.getDiffs(lo, ro);
      d.length.should.equal(1);
      d[0].path.should.equal('a');
      d[0].lObj.should.equal(lo.a);
      d[0].rObj.should.equal(ro.a);
    }); // end of it
  
    it('should stop at a path if options indicate it', function() {
      let lo = {a: {b: {c: 8}}};
      let ro = {a: {b: {c: 7}}};
      let d = utils.getDiffs(lo, ro, undefined, {stopAt: {"a.b": true}});
      d.length.should.equal(1);
      d[0].path.should.equal('a.b');
      d[0].lObj.should.match({c:8});
      d[0].rObj.should.match({c:7});
    }); // end of it
  
    it('should stop at a path of array if options indicate it', function() {
      let lo = {a: {b: [{c: 8}, {c:9}]}};
      let ro = {a: {b: [{c: 8}, {c:10}]}};
      let d = utils.getDiffs(lo, ro, undefined, {stopAt: {"a.b": true}});
      // console.log(d);
      // console.log(d[0].lObj);
      // console.log(d[0].rObj);
      d.length.should.equal(1);
      d[0].path.should.equal('a.b');
      d[0].lObj.should.match([{c: 8}, {c:9}]);
      d[0].rObj.should.match([{c: 8}, {c:10}]);
    }); // end of it
  
    it('should return no path if rObj is null', function() {
      let lo = {a: 1, b: {c: 1}};
      let ro ;
      let d = utils.getDiffs(lo, ro);
      // console.log(d);
      d.length.should.equal(1);
      d[0].path.should.equal('');
      d[0].lObj.should.match({a: 1, b: {c: 1}});
      (ro === undefined).should.equal(true);
    }); // end of it

    it('should use an options compare', function() {
      let lo = {a: 1, b: {z: 0, x: {y: 1}}};
      let ro = {a: 1, b: {z: 0, x: {y: '1'}}};
      let d = utils.getDiffs(lo, ro);
      d.length.should.equal(1);

      let options = {compare: {'b.x.y': (lObj, rObj) => lObj + '' === rObj}};
      d = utils.getDiffs(lo, ro, undefined, options);
      d.length.should.equal(0);
    }); // end of it


  }); // end of diffs describe

});




lib/case.js
/*jslint node: true */
'use strict';

/** 
  * module for eoir case 
  * @module 
  */

/** case data object
  * @typedef {Object} CaseDetails
  * @param {String} eoirId
  * @param {String} aNbr
  * @param {String} aNbrStrip The aNbr stripped of all chars except [0-9]
  * @param {Object} chargingDocDate
  * @param {String} chargingDocDate.orig "mm/dd/yyyy"
  * @param {Date}   chargingDocDate.date  new Date(orig).setUTCHours(18)
  * @param {String} alienName
  * @param {String} hearingLocation
  * @param {String} caseStatus
  * @param {String} clock
  * @param {Array<Hearing>} hearings
  * @param {Array<Proceeding>} proceeding
**/ 

/** hearing data object
  * @typedef {Object} Hearing
  * @param {String} type
  * @param {Object} date
  * @param {String} date.origDate
  * @param {Date}   date.date  new Date(orig).setUTCHours(18)
  * @param {String} date.origTime
  * @param {String} date.timeFrom
  * @param {String} date.timeTo
  * @param {Object} location
  * @param {String} location.addr1
  * @param {String} location.cityStateZip
  * @param {String} iJAssigned

**/

/** proceeding data object
  * @typedef {Object} Proceeding
  * @param {String} sortNbr
  * @param {Object} decisionDate
  * @param {String} decisionDate.orig
  * @param {Date}   decisionDate.date  new Date(orig).setUTCHours(18)
  * @param {String} iJDecision
  * @param {Object} hearingLocation
  * @param {String} hearingLocation.addr1
  * @param {String} hearingLocation.cityStateZip
**/ 

const puppeteer = require('puppeteer');
const _ = require('lodash');
const assert = require('assert'); 

const pfns = require('./pfns.js');
const eoirShared = require('./shared.js');

// const _ = require('lodash');

const selectors = {
  aNbr: 'input#a-number',
    // <input id="a-number" type="text" class="form-control" value="999-888-777" readonly="">
  chargingDocDate: 'input#charging-doc-date',
    // <input id="charging-doc-date" type="text" class="form-control" value="06/27/2016" readonly="">
  alienName: 'input#alien-name',
    // <input id="alien-name" type="text" class="form-control" value="SOMEBODY, NAME " readonly="">
  hearingLocation: 'textarea#hearing-location',
    // <textarea id="hearing-location" class="form-control" rows="5" readonly="">
    // IMMIGRATION COURT
    // 1 FEDERAL DRIVE, SUITE 1850
    // FORT SNELLING, MN 55111
    // (612) 725-3765
    // </textarea>
  caseStatus: 'input#case-status',
    // <input id="case-status" type="text" class="form-control" value="Pending" readonly="">
  clock: 'input#clock',
    // <input id="clock" type="text" class="form-control" value="No Clock" readonly="">
  proceedings: {
    selector: 'table#case-detail-proceedings-table tr',
    sortNbr: 'td:nth-child(2)', 
    decisionDate: 'td:nth-child(3)', 
    iJDecision: 'td:nth-child(4)', 
    hearingLocationAddr1: 'td:nth-child(5) div:nth-child(1)',
    hearingLocationCityStateZip: 'td:nth-child(5) div:nth-child(2)'
      // <td>
      //     <div>566 VETERAN DRIVE., SUITE 101</div>
      //     <div>PEARSALL, TX 78061</div>                
      // </td>
  },
  hearings: {
    selector: 'table#case-detail-future-hearings-table tr',
      //<table id="case-detail-future-hearings-table" class="table dataTable no-footer" role="grid">
    type: 'td:nth-child(1)', 
    date: 'td:nth-child(2)', 
    time: 'td:nth-child(2) span', 
      //<td>03/23/2020 <span style="white-space:nowrap">(01:00 PM - 05:00 PM)</span></td>
    locationAddr1: 'td:nth-child(3) div:nth-child(1)',
    locationCityStateZip: 'td:nth-child(3) div:nth-child(2)',
    iJAssigned: 'td:nth-child(4)', 

  }
};

const caseDetailScrapes = [
  {path: 'aNbr',
   selector: selectors.aNbr,
   getValue: (elH) => elH.value 
  },
  {path: 'aNbrStrip',
   selector: selectors.aNbr,
   getValue: (elH) => elH.value,
   transformValue: (val) => val.replace(/[^\d]/g,'')
  },
  {path: 'chargingDocDate.orig',
   selector: selectors.chargingDocDate,
   getValue: (elH) => elH.value 
  },
  {path: 'chargingDocDate.date',
   selector: selectors.chargingDocDate,
   getValue: (elH) => elH.value,
   transformValue: toDateUTC18
  },
  {path: 'alienName',
   selector: selectors.alienName,
   getValue: (elH) => elH.value 
  },
  {path: 'hearingLocation',
   selector: selectors.hearingLocation,
   getValue: (elH) => elH.value 
  },
  {path: 'caseStatus',
   selector: selectors.caseStatus,
   getValue: (elH) => elH.value 
  },
  {path: 'clock',
   selector: selectors.clock,
   getValue: (elH) => elH.value 
  },

  /********* proceedings ****************/
  {path: 'proceedings[].sortNbr',
   selector: selectors.proceedings.selector + ' ' +
      selectors.proceedings.sortNbr,
   getValue: (elH) => elH.innerHTML
  },

  {path: 'proceedings[].decisionDate.orig',
   selector: selectors.proceedings.selector + ' ' + 
      selectors.proceedings.decisionDate,
   getValue: (elH) => elH.innerHTML
  },
  {path: 'proceedings[].decisionDate.date',
   selector: selectors.proceedings.selector + ' ' + 
      selectors.proceedings.decisionDate,
   getValue: (elH) => elH.innerHTML,
   transformValue: toDateUTC18
  },

  {path: 'proceedings[].iJDecision',
   selector: selectors.proceedings.selector + ' ' + 
      selectors.proceedings.iJDecision,
   getValue: (elH) => elH.innerHTML
  },
  {path: 'proceedings[].hearingLocation.addr1',
   selector: selectors.proceedings.selector + ' ' + 
      selectors.proceedings.hearingLocationAddr1,
   getValue: (elH) => elH.innerHTML
  },
  {path: 'proceedings[].hearingLocation.cityStateZip',
   selector: selectors.proceedings.selector + ' ' + 
      selectors.proceedings.hearingLocationCityStateZip,
   getValue: (elH) => elH.innerHTML
  },

  /********* hearings ****************/
  {path: 'hearings[].type',
   selector: selectors.hearings.selector + ' ' +selectors.hearings.type,
   getValue: (elH) => elH.innerHTML
  },
  {path: 'hearings[].date.origDate',
   selector: selectors.hearings.selector + ' ' +selectors.hearings.date,
   getValue: (elH) => elH.innerHTML.replace(new RegExp("\ <span.*$"), '')
  },
  {path: 'hearings[].date.date',
   selector: selectors.hearings.selector + ' ' +selectors.hearings.date,
   getValue: (elH) => elH.innerHTML.replace(new RegExp("\ <span.*$"), ''),
   transformValue: toDateUTC18
  },
  {path: 'hearings[].date.origTime',
   selector: selectors.hearings.selector + ' ' + selectors.hearings.time,
   getValue: (elH) => elH.innerHTML
  },
  {path: 'hearings[].date.timeFrom',
   selector: selectors.hearings.selector + ' ' + selectors.hearings.time,
   getValue: (elH) => elH.innerHTML.substr(1, 8)
  },
  {path: 'hearings[].date.timeTo',
   selector: selectors.hearings.selector + ' ' + selectors.hearings.time,
   getValue: (elH) => elH.innerHTML.substr(12, 8)
  },
  {path: 'hearings[].location.addr1',
   selector: selectors.hearings.selector + ' ' + 
      selectors.hearings.locationAddr1,
   getValue: (elH) => elH.innerHTML
  },
  {path: 'hearings[].location.cityStateZip',
   selector: selectors.hearings.selector + ' ' + 
      selectors.hearings.locationCityStateZip,
   getValue: (elH) => elH.innerHTML
  },
  {path: 'hearings[].iJAssigned',
   selector: selectors.hearings.selector + ' ' + selectors.hearings.iJAssigned,
   getValue: (elH) => elH.innerHTML
  },

];

module.exports = {
  selectors: selectors, 
  gotoCase: gotoCase,
  getCaseData: getCaseData,
};

/** eoir id - unique id used by eoir to reference a case
  * @typedef {Number} eoirId
**/ 

/** go to case page give eoir id; assumes logged in and disclaimer accepted
  * @param {Page} page
  * @param {eoirId} id
  * @returns {Promise<Response>} 
**/
function gotoCase(page, id) {

  return eoirShared.gotoSigPage(page, 'cases/' + id);

}

/** scrape the data from the case page
  * @param {Page} page
  * @param {String} id
  * @returns {Promise<CaseDetails>}
**/
function getCaseData(page, id) {
  let caseDetails = {}; 
  caseDetails.eoirId = id; 

  return gotoCase(page, id)
  .then(() => pfns.scrapeIt(page, caseDetailScrapes, caseDetails))
  .then(() => caseDetails);
}

/** return a Date object setting UTC Hours to 18
  * @param {String} ds The date in mm/dd/yyyy format
  * @returns {Date | null}
**/ 
function toDateUTC18(ds) {
  let d = new Date(ds).setUTCHours(18); 
  return d ? new Date(d) : null; 
}


lib/home.js
/*jslint node: true */
'use strict';

/** 
  * module for eoir home page
  * @module 
  */

const puppeteer = require('puppeteer');
const pfns = require('./pfns.js');

const eoirHomeUrl = 'https://portale.eoir.justice.gov';
const _ = require('lodash');

const selectors = {
  myCasesLink: [
    'main_frame',   // https://portale.eoir.justice.gov/uniquesig0fac0604b30fa499a20237a1b2f62d54/uniquesig0/SecureUserPortalPortalHomePage/MainFrame.aspx
    'main_iframe',  // https://portale.eoir.justice.gov/uniquesig0fac0604b30fa499a20237a1b2f62d54/uniquesig0/SecureUserPortalPortalHomePage/ContentFrame.aspx
    // '#TilesViewApplicationList1_appDataList_ctl00_appItemLink']
    'a#TilesViewApplicationList1_appDataList_ctl00_appItemLink'],
        // <a href="https://portale.eoir.justice.gov/uniquesig340d161e9d8c4d852c2fb0c49a9d9788/uniquesig0/" 
        //    id="TilesViewApplicationList1_appDataList_ctl00_appItemLink" class="TilesView_AppItemLink" 
        //    target="main_iframe">Electronic Case Information</a>
};

module.exports = {
  selectors: selectors, 
  gotoHome: gotoHome,
};

/** go to portal home page https://portale.eoir.justice.gov; wb login page if user is not logged in
  * @param {Page} page
  * @returns {Promise<Page|Error<Response>>} returns Response if status !== 200
**/ 
function gotoHome(page) {
  // console.log('gotoHome()');
  return page.goto(eoirHomeUrl, {waitUntil: 'networkidle'})
  .then(pfns.checkValidResponse)
  .then(() => {
    return page; 
  });

}
  


lib/login.js
/*jslint node: true */
'use strict';

/** 
  * module for eoir login and basic navigation
  * @module 
  */

const puppeteer = require('puppeteer');
const pfns = require('./pfns.js');
const eoirHome = require('./home.js');
const caseList = require('./case-list.js');

const _ = require('lodash');

const selectors = {
  loginIdInput : 'input#user_name',
    // page.mainFrame: <input class="paramTextbox" type="text" id="user_name" name="user_name"...
  loginPwInput : "input#password",
    // page.mainFrame: <input class="paramTextbox" type="password" id="password" name="password"...
  loginSubmitButton: "input#submit_button",
    // <input border="0" class="button" type="submit" id="submit_button" value="Login">
  // logoffButton: 'a .ctl00_toolBar1_logoutMenu_1'
};

module.exports = {
  selectors: selectors, 
  login: login,
  isLoggedIn: isLoggedIn,
  logoff: logoff
};

/** eoir login credentials (id, pw) object
  * @typedef {Object} LoginCreds
  * @param {String} id
  * @param {String} pw
**/ 


/** log into eoir site; assumes user is logged off
  * @param {Page} page
  * @param {LoginCreds} 
  * @returns {Promise<Page|Error>} Throws error if login not successful
**/ 
function login(page, loginCreds) {
  // console.log('login');

  return eoirHome.gotoHome(page)
  .then(() => {
    return pfns.enterValue(page, pfns.getEl(page, selectors.loginIdInput), loginCreds.id);
  })
  .then(() => {
    return pfns.enterValue(page, pfns.getEl(page, selectors.loginPwInput), loginCreds.pw);
  })
  .then(() => {return pfns.getEl(page, selectors.loginSubmitButton);})
  .then((submitEl) => {
    return submitEl.click();
  })
  .then(() => {
    return page.waitForNavigation({waitUntil: 'networkidle'});
  })
  .then(() => isLoggedIn(page))
  .then(isLI => {
    if(!isLI) {throw new Error("Login failure");}
  })
  .then(() => {
    // accept disclaimer
    return caseList.gotoMyCases(page);
    // return page; 
  })
  ;

    
}

/** returns true (Promise) if there is a user logged in (no check on which user)
  * @param {Page}
  * @returns {Promise<Boolean|Error>}
**/ 
function isLoggedIn(page) {
  /* <td style="white-space:nowrap;">
      <a class="ctl00_toolBar1_logoutMenu_1 StaticMenuItem ctl00_toolBar1_logoutMenu_3" 
        href="javascript:logoff();" 
        style="border-style:none;font-size:1em;">Log Off </a></td>
  */
  // console.log('isLoggedIn');
  return eoirHome.gotoHome(page)
  .then(() => {
    return page.waitForSelector(selectors.loginIdInput, {timeout: 5000});
  })
  .then(() => {
    // console.log('isLoggedIn then');
    return false; 
  }) 
  .catch((e)=> {
    // console.log('isLoggedIn catch');
    return true;  // login id not found so is logged in
  });

}

/** log user off eoir portal 
**/ 
function logoff() {

  // 1. goto home

  // 2. click Log Off if isLoggedIn()

}


lib/shared.js
/*jslint node: true */
'use strict';

/** 
  * module for shared eoir functions
  * @module 
  */

const puppeteer = require('puppeteer');
const pfns = require('./pfns.js');

// const _ = require('lodash');

const eoirBaseUrl = 'https://portale.eoir.justice.gov';

const selectors = {
};

module.exports = {
  getBaseHrefSig: getBaseHrefSig,
  getBaseSigUrl: getBaseSigUrl,
  gotoSigPage: gotoSigPage,
};


/** returns eoir href through unique sig
  * https://portale.eoir.justice.gov/uniquesig340d161e9d8c4d852c2fb0c489ab4948/uniquesig0/
  * @param {String} href
  * @returns {String}
**/
function getBaseHrefSig(href) {
  return href.substr(0, href.indexOf('/uniquesig0')) + '/uniquesig0';
}

/** returns eoir base url through unique sig using the current page url; s/b logged in and accepted disclaimer
  * https://portale.eoir.justice.gov/uniquesig340d161e9d8c4d852c2fb0c489ab4948/uniquesig0/
  * @param {Page} page
  * @returns {String}
  * @example
  *   // returns 'https://portale.eoir.justice.gov/uniquesig340d161e9d8c4d852c2fb0c489ab4948/uniquesig0'
  *   getBaseSigUrl(page);
**/
function getBaseSigUrl(page) {
  return getBaseHrefSig(page.url());
}

/** page.goto(getBaseSigUrl + href)
  * @param {Page} page
  * @param {String} href Such as '/cases' or '/cases/7153245'
  * @returns {Promise} resolves when networkidle
**/ 
function gotoSigPage(page, href) {
  href = href[0] === '/' ? href : '/' + href;
  return page.goto(getBaseSigUrl(page)+ href, {waitUntil: 'networkidle'});
}


lib/pfns.js
/*jslint node: true */
'use strict';

/** 
  * module for puppeteer add-on functions for eoir
  * @module 
  */

const puppeteer = require('puppeteer');
const EventEmitter = require('events');
const _ = require('lodash');

/** Page class as defined by puppeteer 
  * https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#class-page
  * @typedef {Class} Page
**/ 


module.exports = {
  getNewBrowser: getNewBrowser,
  getNewBrowserPage: getNewBrowserPage,
  getEl: getEl,
  gotoFrame: gotoFrame,
  checkValidResponse: checkValidResponse, 
  enterValue: enterValue,
  waitForIdle: waitForIdle,
  clickAndWait: clickAndWait,
  select: select,
  isChecked: isChecked,
  scrapeIt: scrapeIt
};

/** = puppeteer.launch; returns browser instance
  * @see {@link https://github.com/GoogleChrome/puppeteer/blob/v0.11.0/docs/api.md#puppeteerlaunchoptions}
  * 
**/
function getNewBrowser(options) {
  return puppeteer.launch(options);
}


/** return a new browser page
  * @param {Object} options - options object as defined by https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#puppeteerlaunchoptions
  * @returns {Promise<Page>}
**/ 
function getNewBrowserPage(options) {
  return puppeteer.launch(options)
  .then((browser) => {
    return browser.newPage();
  });
}

/** check valid response; throw error with response if note
  * @param {Response} response
  * @returns {Response|Error<response>}
**/ 
function checkValidResponse(response) {
  if(response.status !== 200) {
    throw new Error(response);
  } else {
    return; 
  }
}

/** enter a value in an input element
  * @param {Page} page
  * @param {Promise} elP - Promise that resolves to an element handler
  * @param {String|Number} value - value to enter 
  * @returns {Promise<Page|Error<Response>>}
**/ 
function enterValue(page, elP, value) {
  // console.log(elP);
  return elP
  .then((inputEl) => {
    return inputEl.click();
  })
  .then(() => {
    return page.type(value);
  })
  .then(() => {
    return page; 
  }); 

}

/** goes through all frames in page and returns frame with name requested
  * @param {page}
  * @param {String} frameName
  * @returns {Frame|foundFrame}
**/ 
function getFrame(page, frameName) {
  // console.log('getFrame ' + frameName);
  var foundFrame; 

  dumpFrameTree(page.mainFrame(), '');

  function dumpFrameTree(frame, indent) {
    // console.log(indent + frame.name() + '  : ' + frame.url());
    if(frame.name().trim() === frameName) {
      foundFrame = frame; 
    } else {
      for (let child of frame.childFrames()) {
        dumpFrameTree(child, indent + '  ');
      }
    }
  }

  return foundFrame; 
}

/** get url of frame and goto that url
  * @param {page} page
  * @param {String | Array<String>} frames; frame name or array of frame names
  * @param {Promise|undefined} p - a promise to wait for
  * @returns <Promise>
**/ 
function gotoFrame(page, frames, p) {
  frames = _.isArray(frames) && frames.length === 1 ? frames[0] : frames; 
  p = p ? p : new Promise((resolve) => {resolve();});
  let frameNameToGet = _.isString(frames) ? frames : frames[0];
  let frame = getFrame(page, frameNameToGet);
  let frameUrl = frame.url();
  // console.log(frameNameToGet);
  // console.log(frameUrl);

  if(_.isString(frames)) {
    return p
    .then(() => {
      return page.goto(frameUrl, {waitUntil: 'networkidle'});
    });
  } else {
    let l = p.then(() => { 
        return page.goto(frameUrl, {waitUntil: 'networkidle'});
      });
    frames.splice(0,1);
    return gotoFrame(page, frames, l);
  }
}

/** retrieve an element handler given either a selector string or a selector obj
  * @param {Page} page
  * @param {String|Array.<String>} selectors - if array, then in order of frame name with last being elementselector string
  * @returns {Promise<ElementHandler>}
  * @example 
  *  // returns Promise resolving to element #myelid in main_frame in main_iframe
  *  getEl(page, ['main_frame', 'main_iframe', '#myelid']) 
**/ 
function getEl(page, selectors) {
  // EXAMPLE getEl(page, ['main_frame', 'main_iframe', '#myelid'])

  // cut last string off array and assign as elSelector if array eg '#myelid'
  let elSelector = _.isString(selectors) ? selectors : selectors.splice(-1);  

  // if still elements in array, then those are the frame names; eg ['main_frame', 'main_iframe']
  let frameNames = _.isArray(selectors) && selectors.length > 0 ? 
    selectors : undefined; 

  if(frameNames) {
    return gotoFrame(page, frameNames)
    .then(() => {
      return page.$(elSelector);
    });
  } else {
    return new Promise((resolve) => {
      resolve(page.$(elSelector));
    });
  }

}

/** runs puppeteer page.click and then waits for network idle
  @param {Page} page
  @param {String} selector
  @returns {Promise}
**/ 
function clickAndWait(page, selector) {
  // console.log(selector);
  return page.click(selector)
  .then(() => {return waitForIdle(page);});
}

/** wait for page to network idle
  @param {Page} page
  @returns {Promise}
**/ 
function waitForIdle(page) {
  // console.log('waitForIdle');
  return page.waitForNavigation({waitUntil: 'networkidle'}); 
}

/** select an item in an html select dropdown
  * @see {@link https://stackoverflow.com/questions/45864516/how-to-select-an-option-from-dropdown-select}
  * @todo Replace with page.select when Puppeteer v.12 is released
  * @param {Page} page
  * @param {String} selector
  * @param {!Array<string>} values
  * @returns {Promise}
**/ 
function select(page, selector, ...values) {
  return page.$eval(selector, (element, values) => {
    if (element.nodeName.toLowerCase() !== 'select')
      throw new Error('Element is not a <select> element.');

    const options = Array.from(element.options);

    if (element.multiple) {
      for (const option of options)
        option.selected = values.includes(option.value);
    } else {
      element.value = values.shift();
    }

    element.dispatchEvent(new Event('change'));
    element.dispatchEvent(new Event('input'));
  }, values);
}

/** return if the input radio is selected
  * @param{Page} page
  * @param{String} selector - input selector (without :checked)
  * @returns{Promise<Boolean>}
  * 
**/ 
function isChecked(page, selector) {
  let checkedSelector = selector + ':checked';
  return page.$(checkedSelector)
  .then((el) => {
    return el === null ? false : true;
  });
}

/** 
  * @typedef {Array<Scrape>} Scrapes
**/ 

/** 
  * @typedef {Object} Scrape
  * @param {Selector} selector
  * @param {String} path Path in dot notation. '[]' in path will be replaced with [idx] 
  * @param {Function} getValue Function that takes an ElementHandle found at selector and resolves to value to be assigned to path
  * @param {Function|undefined} transformValue Function to process result of getValue
**/ 

/** scrape data from page into object using Scrape
  * @param {Page} page
  * @param {Scrapes} scrapes
  * @param {Object|undefined} ro The object to be updated and returned
  * @returns {Promise<Object>} returns updated ro Object with props filled in with values
  * @example Non array path
  *  // returns Promise resolving to {gottaName: "Johnny"}
  *  scrapeIt(page, [{selector: 'input#name', path:'date', getValue: (elH) => elH.value}])
  * @example With transformValue
  *  // returns Promise resolving to {date: 1975-05-01T05:00:00.000Z}
  *  scrapeIt(page, [
  *   {selector: 'div#dateStr', 
  *     path:'dateOf', 
  *     getValue: (elH) => elH.value
  *     transformValue: (val) => new Date(val)}])
  * @example Array path
  *  // returns Promise resolving to {names:[{gottaName: "Johnny"}, {gottaName: "Jarvia"}]}
  *  scrapeIt(page, [{selector: 'input#names tr td', path:'names[].gottaName', getValue: (elH) => elH.innerHTML}])
  *  
**/ 
function scrapeIt(page, scrapes, ro={}) {
  var ps = [];

  scrapes.forEach((scrape) => {
    if(scrape.path.indexOf('[]') > -1) {
      ps.push(scrapeIt$$(page, scrape, ro));
    } else { 
      ps.push(scrapeIt$(page, scrape, ro));
    }
  });

  return Promise.all(ps)
  .then(() => ro);
}

/** scrape one element and assign to object following 
  * @param {Page} page
  * @param {Scrape} scrape
  * @param {Object|undefined} ro
  * @returns {Promise<Object|Error>}
**/ 
function scrapeIt$(page, scrape, ro={}) {
  return page.$(scrape.selector)
  .then((elH) => page.evaluate(scrape.getValue, elH))
  .then((val) => scrape.transformValue ? scrape.transformValue(val) : val)
  .then((val) => _.set(ro, scrape.path, val)); 
}

/** scrape multiple ($$) elements and assign to object in array
  * @param {Page} page
  * @param {Scrape} scrape
  * @param {Object|undefined} ro
  * @returns {Promise<Object|Error>}
**/ 
function scrapeIt$$(page, scrape, ro={}) {
  // eg path: "proceedings[].nbr" =>  {proceedings: [{nbr:0}, {nbr:1}]}
  return page.$$(scrape.selector)
  .then((elHs) => {
    let ps = [];
    elHs.forEach((elH, idx) => {
      let path = scrape.path.replace('[]','[' + idx + ']'); // eg "proceedings[0].nbr"
      ps.push(
        page.evaluate(scrape.getValue, elH)
        .then((val) => scrape.transformValue ? scrape.transformValue(val) : val)
        .then((val) => _.set(ro, path, val))
      );
    });
    return Promise.all(ps);
  });
}
